### DISCUSSION TITLE
Issue with KOReader or with the ebooks I'm reading (related to the "align most text left" option)?

### DISCUSSION THREAD

**Discussion by Chairzard** [koreader/koreader#12891] [Q&A] [answered]
I prefer aligning my text to the left while reading rather than justifying it to get consistent spacing between words. My understanding is that the "align most text left" should align text in bodies to the left, but leave everything else untouched. However, in the ebooks I've been reading that have been sourced from Standard Ebooks in particular, toggling this setting also is aligning several elements of the books to the left that should remain centered. Other ebook apps I've used don't do this when text is set to align left (Apple's Books app, Kobo's Nickel, Foliate, etc). I'm not sure if this is a potential bug with the "align most text left" option in KOReader (in which case I can open a bug report) or an issue with the ebooks I'm sourcing (in which case I can reach out to Standard ebooks); I am unable to read HTML well so I'm unsure and any guidance would be appreciated! 

I'm uploading the epubs direct to my ereader, not passing them through calibre or anything else that might edit the file. I've tried switching the style from epub.css to html5.css in the Style settings as well, but this does nothing.

Link to sample ebook that has this issue (public domain in the US; I use the Compatible epub version but it happens in the advanced epub too): https://standardebooks.org/ebooks/h-p-lovecraft/the-case-of-charles-dexter-ward

Parts of the book that are aligned left that should not be with 
1. Setting in question ("align most text left")
![Screenshot_20241211_000050](https://github.com/user-attachments/assets/f8197781-0965-45dc-932c-7d87c1e35768)
2. Bug or formatting issue with chapter titles (area in red should be centered; note how the "II" stays centered correctly):
![Screenshot_20241210_235903](https://github.com/user-attachments/assets/ce9eda02-120a-481c-a3fc-ffa05011084c)
3. Bug or formatting issue in the Colophon (area in red should be centered):
![Screenshot_20241211_001933](https://github.com/user-attachments/assets/08f0b724-c02d-4bc5-9810-5a12d00ec9e8)
4. HTML for screenshot 2.
![Screenshot_20241211_005051](https://github.com/user-attachments/assets/2ab94b64-d3bf-41a8-9e77-5f9dfd24ebb7)
5. HTML for screenshot 3.
![Screenshot_20241211_005410](https://github.com/user-attachments/assets/4a60f060-34a3-4a1c-9753-50e8a17d2ae3)

**Comment by Frenzie** [comment]
The style tweak does mean most text, which includes headers. You can create a custom style tweak based on that one that only affects `p` for example.

**Reply by Chairzard** [reply]
I tinkered around and figured out a solution that works, thank you for pointing me in the right direction!

The default "align most text left" option is: `body, p, li { text-align: left ! important; }`
The custom CSS that I can create that results in text aligned as expected is `body, li { text-align: left ! important; }`

Removing the `p` results in text aligned to the left like one would find with most other ebook apps (Apple Books, Kobo, etc) and is what most users expect with that option; leaving the p in is too aggressive and aligns elements most users would not expect. Do you think requesting removing the `p` from the default "align most text left" option would be accepted as a feature improvement? If so, I can create a request.

**Reply by poire-z** [reply]
> Do you think requesting removing the `p` from the default "align most text left" option would be accepted as a feature improvement? If so, I can create a request.

Not really.
There are multiple solutions to ensure "left align main-content text".
We choose to target `<p>` as these are paragraphs, and should contain the main text.
Your book is a bit crappy, as it puts _some_ heading in `<h2>` and some other part of the heading in a `<p>` (even if `<p epub:type="title">`, we don't exclude these, not even sure it's a commonly thing). If `II` is a H2, it's chapter title shoul be in the same H2 (or another one).

I don't know how other apps handle this. May be they do just as you ended up doing: targeting only `<body>`, and _hope_ that the `text-align:left` will be inherited by its contained `<p>` (and not by the `<h2>`), meaning all of these should have or should not have a proper `text-align: center` or `text-align:justify` to stop the inheritance where it is not needed.
This might work on some books (like this one), but not on others. If the books has `p { text-align: justify }` your `body {text-align: left !important}` won't be able to override the justification of `<p>`.

Anyway, crappy book, our style tweaks not working => Book style tweak is the solution to achieve what you want. Easier now that you learned HTML and CSS in 5 minutes :)

**Reply by Frenzie** [reply]
> Do you think requesting removing the p from the default "align most text left" option would be accepted as a feature improvement?

No, `<p>` is by far the most important thing to have in there. `<body>` is merely there as a kind of backup.

**Reply by mergen3107** [reply]
You can always highlight text in KOReader and use "View HTML" to see exact tags you need.

Then I usually did something global (p) to justify, and then added specific ones (h1 p, h2 p, etc) to have them centered with "!important".

**Reply by Chairzard** [reply, marked as answer]
**EDIT 12/13/2024: Editing this to let all who come across this message know that I've found the best solution to all of this and my old solution is messier. It looks like Standard Ebooks styles their books for web browser engine based readers. Switching to the "HTML5 standard rendering (html5.css)" automatically aligns the text to the left as it usually is in their books and addresses the other formatting issues in one fell swoop; this is the best solution to the issue if you want to align text left. However, if you're looking to justify the text, just use the default "Traditional book look (epub.css)" option. If for some reason you want to use the traditional book look but align text to the left or want to use HTML5 but justify text, you'll need to use custom CSS, as outlined below** 

My original message is below:

> You can always highlight text in KOReader and use "View HTML" to see exact tags you need.
> 
> Then I usually did something global (p) to justify) and then added specific ones (h1 p, h2 p, etc) to have them centered with "!important".

Thank you for this, this was the silver bullet I needed to solve this! Just in case anybody else in the future is trying to read a Standard Ebooks book with text aligned to the left, this is the custom CSS I used to get everything to display properly:
```
body, p, li { text-align: left ! important; }

hgroup p, section.epub-type-colophon p { text-align: center ! important; }
```
Alternatively, you can leave out the first line and enable "align most text left" in settings and just use the second line.

**Comment by Chairzard** [comment]
Thank you both again for the help and clarification (and thank you both for the crash-course in css/HTML)! I'm loving KOReader so far and the custom CSS saved my sanity!
